File: 프로그래머스/1/340213. ［PCCP 기출문제］ 1번 ／ 동영상 재생기/［PCCP 기출문제］ 1번 ／ 동영상 재생기.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int sToT(string time)
{
    string m = time.substr(0, 2);
    string s = time.substr(3, 2);
     
    return stoi(m) * 60 + stoi(s);
}

string tToS(int time)
{
    int m = time / 60;
    int s = time % 60;
    
    string sm = m < 10 ? "0" + to_string(m) : to_string(m);
    string ss = s < 10 ? "0" + to_string(s) : to_string(s);
    return sm + ":" + ss;
}

string solution(string video_len, string pos, string op_start, string op_end, vector<string> commands) {
    string answer = "";
    
    int i_video_len = sToT(video_len);
    int i_pos = sToT(pos);
    int i_op_start = sToT(op_start);
    int i_op_end = sToT(op_end);
    
    for(int i = 0; i < commands.size(); i++)
    {
        // 최소최대 제한, 오프닝 스킵 체크
        if (i_pos < 0) i_pos = 0;
        if (i_video_len < i_pos) i_pos = i_video_len; 
        if (i_op_start <= i_pos && i_pos <= i_op_end) i_pos = i_op_end;
                  
        if (commands[i] == "prev") i_pos -= 10;
        if (commands[i] == "next") i_pos += 10;                     
    }
    
    // 최소최대 제한, 오프닝 스킵 체크
    if (i_pos < 0) i_pos = 0;
    if (i_video_len < i_pos) i_pos = i_video_len; 
    if (i_op_start <= i_pos && i_pos <= i_op_end) i_pos = i_op_end; 
    
    return tToS(i_pos);
}

File: 프로그래머스/1/340213. ［PCCP 기출문제］ 1번 ／ 동영상 재생기/［PCCP 기출문제］ 1번 ／ 동영상 재생기.py
def time_to_seconds(time_str):
    """mm:ss 형식을 초로 변환"""
    m, s = map(int, time_str.split(':'))
    return m * 60 + s

def seconds_to_time(seconds):
    """초를 mm:ss 형식으로 변환"""
    m = seconds // 60
    s = seconds % 60
    return f"{m:02}:{s:02}"

def adjustTime(curr, amountSec, video_len):
    """현재 시간을 amountSec 만큼 조정"""
    curr_sec = time_to_seconds(curr)
    video_sec = time_to_seconds(video_len)
    
    # 10초 전으로 갈 경우 최소 0초를 유지
    new_time = max(0, curr_sec + amountSec)
    
    # 영상 종료 시간 제한
    new_time = min(new_time, video_sec)
    
    return seconds_to_time(new_time)

def solution(video_len, pos, op_start, op_end, commands):
    for c in commands:
        # 오프닝 구간에 있으면 오프닝 끝으로 이동
        if op_start <= pos <= op_end:
            pos = op_end

        if c == 'prev':
            pos = adjustTime(pos, -10, video_len)
        if c == 'next':
            pos = adjustTime(pos, 10, video_len)

    # 마지막으로 오프닝 구간 체크
    if op_start <= pos <= op_end:
        pos = op_end

    return pos
